controlpanel/kubeapi/oidc_patch.py

```diff
@@ -0,0 +1,121 @@
+import base64
+import datetime
+from datetime import timezone
+import json
+import logging
+import os
+import tempfile
+
+import kubernetes
+from kubernetes.config.kube_config import _is_expired
+import oauthlib.oauth2
+from requests_oauthlib import OAuth2Session
+import yaml
+
+
+log = logging.getLogger(__name__)
+
+
+def load_token(self, provider):
+    if "config" not in provider:
+        return
+
+    parts = provider["config"]["id-token"].split(".")
+
+    if len(parts) != 3:  # Not a valid JWT
+        return None
+
+    padding = (4 - len(parts[1]) % 4) * "="
+
+    jwt_attributes = json.loads(base64.b64decode(parts[1] + padding).decode("utf-8"))
+
+    expire = jwt_attributes.get("exp")
+
+    if (expire is not None) and (
+        _is_expired(datetime.datetime.fromtimestamp(expire, tz=timezone.utc))
+    ):
+        log.debug(provider["config"]["idp-issuer-url"])
+
+        self._refresh_oidc(provider)
+
+        if self._config_persister:
+            self._config_persister(self._config.value)
+
+    self.token = f"Bearer {provider['config']['id-token']}"
+
+    return self.token
```

Comment:
#### 2a. Monkey-patching the Kubernetes Python client
>Lines 19 -46

The fix for this base64 padding issue still hasn't made it into the Python Kubernetes client.

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-8c49a844762f3e9fdeb511f2a34ca23eR52)